src/miner.cpp

```diff
@@ -74,321 +74,385 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
     return nNewTime - nOldTime;
 }
 
-CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& scriptPubKeyIn,bool blockstreamCoreCompatible);
+BlockAssembler::BlockAssembler(const CChainParams& _chainparams)
+    : chainparams(_chainparams)
+{
+    // Largest block you're willing to create:
+    nBlockMaxSize = maxGeneratedBlock;
+    // Core:
+    //nBlockMaxSize = GetArg("-blockmaxsize", DEFAULT_BLOCK_MAX_SIZE);
+    // Limit to between 1K and MAX_BLOCK_SIZE-1K for sanity:
+    //nBlockMaxSize = std::max((unsigned int)1000, std::min((unsigned int)(MAX_BLOCK_SIZE-1000), nBlockMaxSize));
+
+    // Minimum block size you want to create; block will be filled with free transactions
+    // until there are no more or the block reaches this size:
+    nBlockMinSize = GetArg("-blockminsize", DEFAULT_BLOCK_MIN_SIZE);
+    nBlockMinSize = std::min(nBlockMaxSize, nBlockMinSize);
+}
+
+void BlockAssembler::resetBlock(const CScript& scriptPubKeyIn)
+{
+    inBlock.clear();
+
+    nBlockSize = reserveBlockSize(scriptPubKeyIn);  // Core: 1000
+    nBlockSigOps = 100;  // Reserve 100 sigops for miners to use in their coinbase transaction
+
+    // These counters do not include coinbase tx
+    nBlockTx = 0;
+    nFees = 0;
+
+    lastFewTxs = 0;
+    blockFinished = false;
+}
+
+uint64_t BlockAssembler::reserveBlockSize(const CScript& scriptPubKeyIn)
+{
+    CBlockHeader h;
+    uint64_t nHeaderSize, nCoinbaseSize;
+
+    // BU add the proper block size quantity to the actual size
+    nHeaderSize = h.GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION);
+    assert(nHeaderSize == 80);  // BU always 80 bytes
+
+    // This serializes with output value, a fixed-length 8 byte field, of zero and height, a varint in
+    // BIP34, of 0.  Add 4 bytes for to allow a 5-byte height varint, sufficient for thousands of years.
+    nCoinbaseSize = coinbaseTx(scriptPubKeyIn, 0, 0).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION) + 4;
+    // BU005 END
```

Comment:
@ptschip @gandrewstone 
Just a note on the +4 here.  Just tried making it definitely too small, +2 for the next smaller 3-byte varint, and make check fails, so that is a good sign our tests cover this close to the limit.
However making it +3, which I believe is one byte too small, passed the tests.  So it seems the tests aren't detecting reserving 1 byte too little of space, unless I have missed something in this analysis.